SNIP-52.md

```diff
@@ -0,0 +1,653 @@
+```
+fun encryptNotificationData(recipientAddr, channelId, plaintext, env) {
+  // ChaCha20 expects a 96-bit (12 bytes) nonce. we will combine two 12 byte buffers to create nonce
+  let saltBytes := nil
+
+  // depending on which mode the channel operates in
+  if inCounterMode(channelId):
+    // counter reflects the nth notification for the given recipient in the given channel
+    let counter := getCounterFor(recipientAddr, channelId)
+
+    // encode uint64 counter in BE and left-pad with 4 bytes of 0x00 to make 12 bytes
+    saltBytes := concat(zeros(4), uint64BigEndian(counter))
+
+  // otherwise, channel is in TxHash Mode
+  else:
+    // take first 12 bytes of tx hash (make sure to decode the hex string)
+    saltBytes := slice(hexToBytes(env.txHash), 0, 12)
+
+  // take the first 12 bytes of the channel id's sha256 hash
+  let channelIdBytes := slice(sha256(utf8ToBytes(channelId)), 0, 12)
+
+  // produce the nonce by XOR'ing the two previous 12-byte results
+  let nonce := xorBytes(channelIdBytes, saltBytes)
+
+  // right-pad the plaintext with 0x00 bytes until it is of the desired length (keep in mind, payload adds 16 bytes for tag)
+  let message := concat(plaintext, zeros(DATA_LEN - len(plaintext)))
+
+  // construct the additional authenticated data
+  let aad := concatStrings(env.blockHeight, ":", env.txHash)
+
+  // encrypt notification data for this event
+  let seed := getSeedFor(recipientAddr)
+  let [ciphertext, tag] := chacha20poly1305_encrypt(key=seed, nonce=nonce, message=message, aad=aad)
+
+  // concatenate ciphertext and 16 bytes of tag (note: crypto libs typically default to doing it this way in `seal`)
+  let payload := concat(ciphertext, tag)
+
+  return payload
+}
+```
+
+
+Pseudocode for decrypting data from notifications (client):
+```
+fun decryptNotificationData(contractAddr, channelId, payload, env) {
+  // depending on which mode the channel operates in
+  if inCounterMode(channelId):
+    // counter reflects the nth notification for the given recipient in the given channel
+    let counter := getCounterFor(recipientAddr, channelId)
+
+    // encode uint64 counter in BE and left-pad with 4 bytes of 0x00 to make 12 bytes
+    saltBytes := concat(zeros(4), uint64BigEndian(counter))
+
+  // otherwise, channel is in TxHash Mode
+  else:
+    // take first 12 bytes of tx hash (make sure to decode the hex string)
+    saltBytes := slice(hexToBytes(env.txHash), 0, 12)
+
+  // ChaCha20 expects a 96-bit (12 bytes) nonce
+  // take the first 12 bytes of the channel id's sha256 hash
+  let channelIdBytes := slice(sha256(utf8ToBytes(channelId)), 0, 12)
+
+  // produce the nonce by XOR'ing the two previous 12-byte results
+  let nonce := xorBytes(channelIdBytes, counterBytes)
+
+  // construct the additional authenticated data
+  let aad := concatStrings(env.blockHeight, ":", env.txHash)
+
+  // split payload
+  let ciphertext := slice(payload, len(payload) - 16)
+  let tag := slice(payload, len(ciphertext))
```

Comment:
here I think there's an error as it seems that the `tag` and the `ciphertext` are receiving the same part of the payload, the tag should be the second element of the slice

Reply:
good catch 👌